File: cpp/tests/spin_lock_test.cpp
#include "test_harness.hpp"
#include "spin_lock.hpp"
#include <thread>
#include <vector>

int main() {
    TestHarness t;
    std::cout << "SpinLockTest\n";

    SpinLock lock;
    constexpr int THREADS = 8;
    constexpr int PER_THREAD = 20000;

    int counter = 0;
    int insideCount = 0;
    bool mutualExclusionViolated = false;

    std::vector<std::thread> threads;
    for (int i = 0; i < THREADS; ++i) {
        threads.emplace_back([&] {
            for (int j = 0; j < PER_THREAD; ++j) {
                lock.lock();
                ++insideCount;
                if (insideCount != 1) mutualExclusionViolated = true;
                ++counter; // plain increment -- safe only because of the lock
                --insideCount;
                lock.unlock();
            }
        });
    }
    for (auto& th : threads) th.join();

    t.checkEquals("no lost updates under contention", THREADS * PER_THREAD, counter);
    t.check("mutual exclusion never violated (never >1 thread inside)",
            !mutualExclusionViolated);

    return t.summary("SpinLockTest");
}

File: java/tests/SpinLockTest.java
/** Same suite as cpp/tests/spin_lock_test.cpp -- see that file for the
 *  C++ side of this exact comparison. */
public class SpinLockTest {

    static int run() throws InterruptedException {
        TestHarness t = new TestHarness();
        System.out.println("SpinLockTest");

        SpinLock lock = new SpinLock();

        // Mutual exclusion: an "inside the critical section" counter
        // must never be observed above 1 by another thread that also
        // holds the lock while checking it.
        final int THREADS = 8;
        final int PER_THREAD = 20_000;
        int[] counter = {0};
        int[] insideCount = {0};
        boolean[] mutualExclusionViolated = {false};

        Thread[] threads = new Thread[THREADS];
        for (int i = 0; i < THREADS; i++) {
            threads[i] = new Thread(() -> {
                for (int j = 0; j < PER_THREAD; j++) {
                    lock.lock();
                    try {
                        insideCount[0]++;
                        if (insideCount[0] != 1) mutualExclusionViolated[0] = true;
                        counter[0]++; // plain increment -- safe only because of the lock
                        insideCount[0]--;
                    } finally {
                        lock.unlock();
                    }
                }
            });
            threads[i].start();
        }
        for (Thread th : threads) th.join();

        t.check("no lost updates under contention", counter[0] == THREADS * PER_THREAD);
        t.check("mutual exclusion never violated (never >1 thread inside)",
                !mutualExclusionViolated[0]);

        return t.summary("SpinLockTest");
    }

    public static void main(String[] args) throws InterruptedException {
        System.exit(run());
    }
}
